hub/core/compression/base.py

```diff
@@ -0,0 +1,163 @@
+from typing import Union, Dict
+from abc import ABC, abstractmethod  # type: ignore
+from numcodecs.abc import Codec  # type: ignore
+
+import numpy as np
+from .constants import WEBP_COMPRESSOR_NAME, IMAGE_SHAPE_ERROR_MESSAGE, MSGPACK
+
+
+class BaseNumCodec(ABC):
+    """Base class for numcodec compressors"""
+
+    @abstractmethod
+    def encode(self, input: np.ndarray) -> bytes:
+        pass
+
+    @abstractmethod
+    def decode(self, bytes: bytes) -> Union[np.ndarray, bytes]:
+        pass
+
+    @property
+    @abstractmethod
+    def __name__(self):
+        pass
+
+
+class BaseImgCodec(ABC, Codec):
+    """Base class for image codecs"""
+
+    def __init__(self, single_channel: bool = True) -> None:
+        self.single_channel = single_channel
+
+    @property
+    def __name__(self):
+        raise NotImplementedError()
+
+    def encode(self, arr: np.ndarray) -> bytes:
+        """
+        Encode array with one or multiple images.
+
+        Example:
+            img = np.ones((4, 100, 100, 1))
+            encoded_img = codec.encode(img)
+
+        Args:
+            arr (np.ndarray): Image data to be encoded. Can contain image/s with shapes:
+                (N, M) if image has one channel and single_channel is False,
+                (N, M, 1) if image has one channel and single_channel is True,
+                (N, M, 3). In case of multiple images they should be stacked across the first axis.
```

Comment:
```suggestion
                (N, M, 3) if image has multiple channels. 
                In case of multiple images they should be stacked across the first axis.
```